### PR TITLE
DNS Fix - Now demo works without fixed IPs

### LINKED ISSUE #23: bug: DNS not resolving service name to IP address
Related issue:
1. https://github.com/moby/moby/issues/47403
2. https://github.com/second-state/microservice-rust-mysql/pull/22

The current status:

1. WASM container cannot resolve the service name
2. The DNS server of docker-compose is still `127.0.0.11:53`
3. The `DNS_SERVER` magic is not working.
4. The wasi-socket component inside WasmEdge is not changed.
5. We should check all used wasi-socket crates to see if there are some modifications related to this problem <-- We are here now

### PR DESCRIPTION
**Pull request by aalonsolopez**
Fixes #23

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/). Here are my reviews of code commits in this PR.

------

### Summary:

#### Potential Issues and Errors:
1. **Dependency Management Risks:** The use of Git sources, modifications to dependency paths, and changes in submodule configurations introduce risks related to version control, security vulnerabilities, inconsistent builds, and project maintainability. Addressing these potential problems is crucial to ensure the stability and reliability of the project.
2. **Missing or Unintended Changes:** Several patches highlight missing files, inconsistencies, incomplete context, and unintentional dependency modifications. These issues can lead to functionality errors, deployment challenges, and confusion among developers. Thorough validation and documentation are essential to mitigate these risks.

#### Important Findings:
1. **Dependency Handling:** The changes involve updating dependencies, submodule paths, and Dockerfile configurations. Careful consideration of the necessity, impact, and implications of these modifications is vital to maintain project integrity and facilitate collaboration.
2. **Code Structure and Formatting:** Consistent formatting, clear documentation, and organized project structure are critical aspects highlighted across the patches. Ensuring code readability, maintainability, and adherence to best practices can enhance development efficiency.
3. **Testing and Validation:** Emphasizing thorough testing after introducing changes to dependencies, paths, and configurations is crucial to guarantee the correct functionality of the demo and prevent unexpected issues during deployment and execution.

In conclusion, while the patches aim to address the issue of fixed IPs in the demo, attention to potential problems, validation of changes, documentation of modifications, and comprehensive testing are key considerations to ensure a successful and stable integration of the improvements into the project.

## Details

### [Commit aa0069b0cd41ced24c34a0076d237bd88b0f40c7](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/aa0069b0cd41ced24c34a0076d237bd88b0f40c7)
Key changes:
- The `Cargo.toml` file has been modified to change the `tokio_wasi` dependency from a specific version to a Git source with specified features.

Potential problems:
1. Using a Git source for a dependency can lead to issues with version control and dependency management. It's important to ensure that the Git repository is stable and regularly maintained to avoid unexpected issues.
2. Depending on a Git source can introduce potential security risks if the source is not secure or if it gets compromised. It's essential to verify the authenticity and trustworthiness of the Git repository before depending on it in the project.
3. Switching to a Git source may impact build times and could introduce inconsistencies across different build environments if the Git source is not pinned to a specific commit or tag. It's advisable to use a stable source or tag if possible to maintain consistency.

Overall, while the change may address the issue related to fixed IPs in the demo, the potential problems mentioned above should be considered and addressed before merging this patch. It would be beneficial to review and test the change thoroughly to ensure its stability and maintainability in the project.

### [Commit 9366a9a13f547d7b370165fc61a80e43817fcdc5](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/9366a9a13f547d7b370165fc61a80e43817fcdc5)
**Key Changes:**
1. The patch updates the `Dockerfile` by changing the `COPY` command from `COPY Cargo.toml orders.json update_order.json .` to `COPY Cargo.toml orders.json update_order.json ./`, adding a forward slash at the end of the destination directory.

**Potential Problems:**
1. **Missing Files:** The change in the `COPY` command indicates that the `Dockerfile`, `Cargo.toml`, `orders.json`, and `update_order.json` files are being copied to the root directory in the container. Ensure that these files exist and are essential for the application to function correctly.
2. **Consistency:** It's important to ensure consistent formatting throughout the file. While the change seems minor, it might be indicative of inconsistent practices elsewhere in the codebase. Consider revisiting other parts of the codebase for similar inconsistencies.
3. **Dependency Management:** If any of the files being copied are dependencies, verify if they are being fetched or managed properly in the overall dependency management strategy of the project.

Overall, the patch seems straightforward and mainly deals with a formatting change. However, it's crucial to validate the necessity and implications of copying these specific files to the root directory in the container.

### [Commit 6c9ec51362a5d43d512de64f5e9ed00edbd8a972](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/6c9ec51362a5d43d512de64f5e9ed00edbd8a972)
### Key Changes:
1. Added local references for the dependencies h2, hyper, mysql_async_wasi, and tokio in Cargo.toml.
2. Created new directories for each local dependency: h2, hyper, mysql_async_wasi, and tokio.
3. Added subproject commits for each local dependency in their respective directories.

### Potential Problems:
1. **Incomplete Information**: The patch seems to add local references for certain dependencies but lacks context on why these changes were made. The commit message and description should provide reasoning for these modifications.
   
2. **Dependency Management**: While using local paths for dependencies can be useful for development, it may introduce issues when deploying or sharing the project with others. Ensure that this approach aligns with the project's long-term goals in terms of dependency management.

3. **Lack of Testing**: If these changes impact the overall functionality of the project, thorough testing is required to guarantee that the demo and other functionalities work as intended with the updated dependencies.

4. **Code Maintenance**: The patch introduces new directories for local dependencies. Make sure that the project structure remains organized and clear following these additions to prevent confusion for other developers working on the project.

5. **Possible Inconsistencies**: The use of local paths for dependencies might differ from the standard dependency management practices in the project. Confirm that this deviation does not conflict with established conventions or best practices unless there is a compelling reason for doing so.

Overall, the patch introduces local references for dependencies, which can be beneficial for specific development scenarios. However, it is essential to address the potential problems mentioned above to ensure that these changes do not introduce unexpected issues or hinder the project's overall stability and scalability.

### [Commit 1af0e7415fc3acf13d3f7d41039b78d2ff1d7d76](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/1af0e7415fc3acf13d3f7d41039b78d2ff1d7d76)
Key Changes:
1. Added submodules for `h2`, `hyper`, `tokio`, and `mysql_async_wasi` with their respective paths, URLs, and branches in the `.gitmodules` file.

Potential Problems:
1. The patch is adding submodules, which could increase the complexity of the project and dependencies. Make sure that the addition of these submodules is necessary and does not introduce unwanted dependencies or conflicts.
2. Ensure that the URLs for the submodules are correct and accessible. If any of these URLs are incorrect or change in the future, it could cause issues with fetching the submodules.
3. The patch does not have a newline at the end of the `.gitmodules` file. While this is not a critical issue, it is good practice to have a newline at the end of files for consistency.

### [Commit e6296683b92c7500288549dfae2916724a4f7ebb](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/e6296683b92c7500288549dfae2916724a4f7ebb)
Key changes in the patch:
1. Updated submodule paths in the `.gitmodules` file to use relative paths under `./local_deps`.
2. Updated submodule paths for `h2`, `hyper`, `tokio`, and `mysql_async_wasi`.

Potential problems:
1. The changes seem to be related to updating submodule paths in the `.gitmodules` file. Ensure that these changes align with the project's overall structure and conventions.
2. Verify that the new submodule paths under `./local_deps` are correctly pointing to the respective repositories (`h2`, `hyper`, `tokio`, `mysql_async_wasi`).
3. Check if any dependencies or scripts in the project rely on the previous submodule paths, as updating them could potentially break functionalities.
4. It's always a good practice to test the changes thoroughly after updating submodule paths to ensure the project builds and functions as expected.


### [Commit c094fcff6cded2ba5d2492340f747996c8e6f3a2](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/c094fcff6cded2ba5d2492340f747996c8e6f3a2)
Key Changes:
- This patch removes the submodule dependencies for `h2`, `hyper`, `mysql_async_wasi`, and `tokio`.
- The patch includes deletions of 1 line for each of the mentioned dependencies.

Potential Problems:
- Removing submodule dependencies might lead to compatibility issues with the code that relied on these dependencies.
- If the functionality provided by these submodules is still required, the code may break due to missing dependencies.
- It is essential to ensure that the codebase does not rely on the functionality provided by the deleted submodules.

Overall, the patch seems to be aimed at removing unnecessary dependencies, but it is crucial to verify that the code still functions correctly without these submodules and that no important functionality has been lost.

### [Commit 864e4552a9d45ba680622c5aa4b220c12bc4ef5d](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/864e4552a9d45ba680622c5aa4b220c12bc4ef5d)
Key Changes:
- The paths for the submodules in the .gitmodules file have been modified to remove the leading "./" part.
- The submodule paths have been updated to use relative paths instead of absolute paths.

Potential Problems:
1. One potential issue could arise if other parts of the codebase rely on the submodule paths being absolute. Changing them to relative paths could break those dependencies.
2. The removal of "./" in the submodule paths might affect the submodule resolution mechanism if the project structure relies on the paths starting with "./".
3. It is good practice to include a newline at the end of the file, and the patch currently lacks that.

Overall, the changes seem straightforward, but it is essential to ensure that these modifications do not impact other parts of the project relying on the previous submodule paths.

### [Commit 0533d685bcac55fb5bbe8a658ea00e663c839b98](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/0533d685bcac55fb5bbe8a658ea00e663c839b98)
Key Changes:
1. Added `tokio_wasi` dependency with the path `./local_deps/tokio/tokio`.
2. Reordered the dependencies in the `Cargo.toml` file.

Potential Problems:
1. Reordering dependencies might introduce unexpected behaviors if there are specific dependencies or versions required by the project.
2. It seems like the removal of the `tokio_util_wasi` dependency might be unintentional, as it was present before the patch.
3. The addition of the same dependency `tokio_wasi` twice might be a mistake or could lead to conflicts if they are different versions or configurations.

Overall, the main concern here is the reordering of dependencies and the potential unintended removal or duplication of dependencies. These changes should be reviewed carefully to ensure they do not introduce issues in the project.

### [Commit 7538d0bc437de4b3a60d2f3684b7929e01ccf30d](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/7538d0bc437de4b3a60d2f3684b7929e01ccf30d)
Key Changes:
1. The patch adds a new line to the Dockerfile, specifically `COPY local_deps ./local_deps`. This change copies the contents of the `local_deps` folder into the Docker image during the build process.

Potential Problems:
1. It is important to ensure that the `local_deps` folder exists and contains the necessary dependencies required for the demo to work without fixed IPs. If the folder is missing or does not contain the correct dependencies, the demo may still face issues.
2. The patch does not provide details about the contents or purpose of the `local_deps` folder. It would be beneficial to document this information either in the code comments or in the README to help developers understand its importance and contents.

Overall, the addition of the `local_deps` folder to the Dockerfile seems like a necessary step to make the demo work without fixed IPs, assuming that the folder contains the required dependencies. However, additional information and documentation would improve the clarity and maintainability of the changes.

### [Commit 03b0667b450c283b347313647fdf44bf7fb9b671](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/03b0667b450c283b347313647fdf44bf7fb9b671)
Key Changes:
- The patch rolls back the version of `mysql_async_wasi` from a local dependency to version `0.31`.
- The `Cargo.toml` file was modified, specifically changing the version of `mysql_async_wasi`.

Potential Problems:
1. **Compilation Errors**: The reason for rolling back to version `0.31` of `mysql_async_wasi` is mentioned as compilation errors. It's critical to ensure that by reverting to the previous version, the compilation issues are resolved without introducing new issues. A thorough testing phase is needed to confirm that the codebase now compiles correctly.

2. **Dependency Consistency**: It seems there were multiple instances of `mysql_async_wasi` being defined in the `Cargo.toml` file. It's crucial to ensure that the dependencies are correctly managed and that there are no conflicting or redundant dependencies declared.

Further review of the overall impact of this change and testing the application after applying the patch is necessary to confirm that the rollback effectively resolves the compilation errors without causing regressions or introducing new issues.

### [Commit b267bdb02f5c12ffe5dff36f762dee80d610b13b](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/b267bdb02f5c12ffe5dff36f762dee80d610b13b)
Key Changes:
1. Updated the version of `mysql_async_wasi` to `=0.31.5`.
2. Updated the version of `hyper_wasi` to `"0.15"`.
3. Updated the version of `tokio_wasi` to `"1"` and included additional features - `io-util`, `fs`, `net`, `time`, `rt`, `macros`.

Potential Problems:
1. The removal of the `COPY local_deps` command in the Dockerfile might cause issues if those dependencies are still required for the build process. It would be good to confirm if these dependencies are no longer needed or if there's an alternative method in place.
2. The changes in dependency versions could potentially introduce compatibility issues with other parts of the codebase or with external dependencies. Thorough testing is required to ensure that the application functions as expected with these updated versions.
3. Verify that the specific versions mentioned in the patch are necessary and compatible with the rest of the project's dependencies. It's important to ensure that the specified versions are explicitly required for the project's functionality.

### [Commit 3d4ac2444af4fce54ea3f47e6925b68d0ad91818](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/3d4ac2444af4fce54ea3f47e6925b68d0ad91818)
Key Changes:
- This patch deletes the ".gitmodules" file.
- Removes submodule configurations for "h2", "hyper", "tokio", and "mysql_async_wasi".
  
Potential Problems:
- Deleting the ".gitmodules" file removes the configurations for submodules, which might have been necessary for the project's dependencies.
- If the submodules were crucial for the project's functionality, their removal could break the project.
- There might be dependencies on specific branches of the submodules that are now removed.

Recommendation:
- Check if the submodules are indeed no longer needed for the project.
- If the submodules are still required, consider adding them back or updating their configurations in a different way.
- Ensure that removing the submodules and ".gitmodules" file does not introduce any unforeseen issues or dependencies elsewhere in the project.

### [Commit a730e2a461f31b645c52e15e62aed18ff0dd1ee2](https://github.com/second-state/microservice-rust-mysql/pull/24/commits/a730e2a461f31b645c52e15e62aed18ff0dd1ee2)
Key Changes:
- In the `docker-compose.yml` file, the `MYSQL_ROOT_PASSWORD` value has been changed to `whalehello`.

Potential Problems:
1. No functional changes: The patch seems to make no meaningful functional changes, as it simply updates the `MYSQL_ROOT_PASSWORD` value in the `docker-compose.yml` file.
2. No description: The commit message does not provide any context or explanation for the change, making it harder for reviewers to understand the purpose of the modification.
3. No new line at end of file: The diff shows that there is no new line added at the end of the file, which might not align with the coding standards of the project.
4. No version bump: If this change affects the behavior of the application, it might require a version bump or a more detailed explanation in the commit message to indicate the impact.

**Comment by aalonsolopez** [comment]
Solves #23

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[submodule "local_deps/h2"]
```
Comment (hydai) [inline .gitmodules]: I don't think we need the submodule to fix this issue.

File: Cargo.toml [hunk 2]
```diff
@@ -8,6 +8,6 @@ anyhow = "1.0"
 serde_json = "1.0"
 serde = { version = "1.0", features = ["derive"] }
 url = "2.3"
-mysql_async_wasi = "0.31"
+mysql_async_wasi = "=0.31.5"
 hyper_wasi = { version = "0.15", features = ["full"] }
-tokio_wasi = { version = "1", features = ["io-util", "fs", "net", "time", "rt", "macros"] }
+tokio_wasi = { version = "1", features = ["io-util", "fs", "net", "time", "rt", "macros"] }
```
Comment (hydai) [inline Cargo.toml:13]: Why does this line change?

**Review by hydai — CHANGES_REQUESTED** [review]
We should also modify the fixed IP back to the DNS server.

File: .gitmodules (hunk 1 above)
Comment (aalonsolopez) [reply .gitmodules]: This submodule does not exist anymore on last commit, that is what resolves the problem

File: Cargo.toml (hunk 2 above)
Comment (aalonsolopez) [reply Cargo.toml:13]: Is the same line... So IDK, but it does not seem to matter

**Comment by aalonsolopez** [comment]
> We should also modify the fixed IP back to the DNS server.

So you mean deleting the fixed IP for the DNS server? I'm pretty sure that it didn't work anytime

**Comment by aalonsolopez** [comment]
I mean, I should have premerged the testing branch to only show the relevant change but there are all the changes with all the local deps that I was investigating, which are irrelevant to the final solution.

**Comment by aalonsolopez** [comment]
btw, DNS_SERVER env variable is still necessary

**Comment by aalonsolopez** [comment]
> We should also modify the fixed IP back to the DNS server.

And I don't fully understand this, sry

(sorry for closing the PR, I missclicked)

File: .gitmodules (hunk 1 above)
Comment (hydai) [reply .gitmodules]: It still remains. That's why I can review this file. Please squash your commits.

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by hydai** [comment]
> > We should also modify the fixed IP back to the DNS server.
> 
> So you mean deleting the fixed IP for the DNS server? I'm pretty sure that it didn't work anytime

I got so confused. Your title shows that we don't need the fixed IPs anymore.

**Comment by aalonsolopez** [comment]
> > > We should also modify the fixed IP back to the DNS server.
> > 
> > 
> > So you mean deleting the fixed IP for the DNS server? I'm pretty sure that it didn't work anytime
> 
> I got so confused. Your title shows that we don't need the fixed IPs anymore.

We don't need the fixed IPs for every container as I did on #22, but as we talked on the discord discussion we still need the DNS_SERVER IP

**Comment by hydai** [comment]
> > > > We should also modify the fixed IP back to the DNS server.
> > > 
> > > 
> > > So you mean deleting the fixed IP for the DNS server? I'm pretty sure that it didn't work anytime
> > 
> > 
> > I got so confused. Your title shows that we don't need the fixed IPs anymore.
> 
> We don't need the fixed IPs for every container as I did on #22, but as we talked on the discord discussion we still need the DNS_SERVER IP

Yup. That's why I said we need to modify the docker compose file from FIXED IPs to DNS_SERVER IP.

**Review by hydai — APPROVED** [review]

**Comment by hydai** [comment]
Thanks.

**Pull request merged**